index.js

```diff
@@ -70,20 +72,71 @@ function format(text, opts, addAlignmentSize) {
   return str;
 }
 
-function formatRange(text, opts) {
+function findNodeByOffset(node, offset, opts, text) {
+  const start = util.locStart(node);
+  const end = util.locEnd(node);
+  if (start <= offset && offset <= end) {
+    for (const childNode of comments.getSortedChildNodes(node)) {
+      const childResult = findNodeByOffset(childNode, offset, opts, text);
+      if (childResult) {
+        return childResult;
+      }
+    }
+
+    try {
+      parser.parse(text.slice(start, end), opts);
+      return node;
+    } catch (err) {
+      return;
+    }
+  }
+}
+
+function calculateRange(text, opts, ast) {
+  const startNode = findNodeByOffset(ast, opts.rangeStart, opts, text);
+  const endNode = findNodeByOffset(ast, opts.rangeEnd, opts, text);
+  const rangeStart = Math.min(util.locStart(startNode), util.locStart(endNode));
+  const rangeEnd = Math.max(util.locEnd(startNode), util.locEnd(endNode));
+
+  const rangeString = text.slice(rangeStart, rangeEnd);
+
+  // Try to extend the range backwards to the beginning of the line.
+  // This is so we can detect indentation correctly and restore it.
+  const rangeAst = parser.parse(rangeString, opts);
   // Use `Math.min` since `lastIndexOf` returns 0 when `rangeStart` is 0
-  const rangeStart = Math.min(
-    opts.rangeStart,
-    text.lastIndexOf("\n", opts.rangeStart) + 1
+  const rangeStart2 = Math.min(
+    rangeStart,
+    text.lastIndexOf("\n", rangeStart) + 1
   );
-  // Use `text.length - 1` as the maximum since `indexOf` returns -1 if `fromIndex >= text.length`
-  const fromIndex = Math.min(opts.rangeEnd, text.length - 1);
-  const nextNewLineIndex = text.indexOf("\n", fromIndex);
-  const rangeEnd = (nextNewLineIndex < 0 ? fromIndex : nextNewLineIndex) + 1; // Add one to make rangeEnd exclusive
-
-  if (0 < rangeStart || rangeEnd < text.length) {
-    const rangeString = text.substring(rangeStart, rangeEnd);
-    const alignmentSize = getAlignmentSize(
+  const rangeString2 = text.slice(rangeStart2, rangeEnd);
+  try {
+    const rangeAst2 = parser.parse(rangeString2, opts);
+    if (cleanAST(rangeAst) === cleanAST(rangeAst2)) {
```

Comment:
On second thought, we can still keep this behavior without extra parsing/printing. See https://github.com/prettier/prettier/pull/1659/commits/2da0a03cefe75313b905ac32f8b4c001d43ed849